fun-token-contract/src/lib.rs

```diff
@@ -0,0 +1,596 @@
+            "The new owner should be different from the current owner"
+        );
+        // Retrieving the account from the state.
+        let mut account = self.get_account(&owner_id);
+
+        // Checking and updating unlocked balance
+        if account.balance < amount {
+            env::panic(b"Not enough balance");
+        }
+        account.balance -= amount;
+
+        // If transferring by escrow, need to check and update allowance.
+        let escrow_account_id = env::predecessor_account_id();
+        if escrow_account_id != owner_id {
+            let allowance = account.get_allowance(&escrow_account_id);
+            if allowance < amount {
+                env::panic(b"Not enough allowance");
+            }
+            account.set_allowance(&escrow_account_id, allowance - amount);
+        }
+
+        // Saving the account back to the state.
+        self.set_account(&owner_id, &account);
+
+        // Deposit amount to the new owner and save the new account to the state.
+        let mut new_account = self.get_account(&new_owner_id);
+        new_account.balance += amount;
+        self.set_account(&new_owner_id, &new_account);
+        self.refund_storage(initial_storage);
+    }
+
+    /// Transfer `amount` of tokens from the caller of the contract (`predecessor_id`) to
+    /// `new_owner_id`.
+    /// Act the same was as `transfer_from` with `owner_id` equal to the caller of the contract
+    /// (`predecessor_id`).
+    /// Requirements:
+    /// * Caller of the method has to attach deposit enough to cover storage difference at the
+    ///   fixed storage price defined in the contract.
+    #[payable]
+    pub fn transfer(&mut self, new_owner_id: AccountId, amount: U128) {
+        // NOTE: New owner's Account ID checked in transfer_from.
+        // Storage fees are also refunded in transfer_from.
+        self.transfer_from(env::predecessor_account_id(), new_owner_id, amount);
+    }
+
+    /// Returns total supply of tokens.
+    pub fn get_total_supply(&self) -> U128 {
+        self.total_supply.into()
+    }
+
+    /// Returns balance of the `owner_id` account.
+    pub fn get_balance(&self, owner_id: AccountId) -> U128 {
+        self.get_account(&owner_id).balance.into()
+    }
+
+    /// Returns current allowance of `escrow_account_id` for the account of `owner_id`.
+    ///
+    /// NOTE: Other contracts should not rely on this information, because by the moment a contract
+    /// receives this information, the allowance may already be changed by the owner.
+    /// So this method should only be used on the front-end to see the current allowance.
+    pub fn get_allowance(&self, owner_id: AccountId, escrow_account_id: AccountId) -> U128 {
+        assert!(
+            env::is_valid_account_id(escrow_account_id.as_bytes()),
+            "Escrow account ID is invalid"
+        );
+        self.get_account(&owner_id)
+            .get_allowance(&escrow_account_id)
+            .into()
+    }
+}
```

Comment:
we need one more function here:

```
pub fn mint(&mut self)
```
Which will mint 10 tokens to `env::predecessor_account_id()` and increase the `total_balance`.